cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -611,6 +611,110 @@ class AzureFileSystem::Impl {
     RETURN_NOT_OK(ptr->Init());
     return ptr;
   }
+
+  Status CreateDir(const AzurePath& path) {
+    if (path.container.empty()) {
+      return Status::Invalid("Cannot create an empty container");
+    }
+
+    if (path.path_to_file.empty()) {
+      auto container_client =
+          blob_service_client_->GetBlobContainerClient(path.container);
+      try {
+        auto response = container_client.Create();
+        if (response.Value.Created) {
+          return Status::OK();
+        } else {
+          const auto& body = response.RawResponse->GetBody();
+          std::string_view body_text(reinterpret_cast<const char*>(body.data()),
+                                     body.size());
+          return Status::IOError("Failed to create a container: ", path.container, " (",
+                                 container_client.GetUrl(),
+                                 "): ", response.RawResponse->GetReasonPhrase(), " (",
+                                 static_cast<int>(response.RawResponse->GetStatusCode()),
+                                 "): ", body_text);
+        }
+      } catch (const Azure::Storage::StorageException& exception) {
+        return internal::ExceptionToStatus(
+            "Failed to create a container: " + path.container + " (" +
+                container_client.GetUrl() + ")",
+            exception);
+      }
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto hierarchical_namespace_enabled,
+                          hierarchical_namespace_.Enabled(path.container));
+    if (!hierarchical_namespace_enabled) {
+      return Status::NotImplemented(
+          "Cannot create a directory without hierarchical namespace: ", path.full_path);
+    }
+    auto directory_client = datalake_service_client_->GetFileSystemClient(path.container)
+                                .GetDirectoryClient(path.path_to_file);
```

Comment:
I can see why `path_to_file` is actually the name of the directory in this context, but maybe a different name for this struct would make things less confusing? This segment of filesystem paths is usually called "basename" [1].

[1] https://en.wikipedia.org/wiki/Basename

Reply:
I spent quite some time trying to figure out what `if (path.path_to_file.empty()) {` meant here.

`path.basename.empty()` would be more clear IMO.

cc @Tom-Newton

Reply:
I think that "basename" isn't suitable for this case.
I think that "d" is the basename of "a/b/c/d" but `path_to_file` is "b/c/d". ("a" is `container`.)

I think that `path` is suitable for "b/c/d" but `AzurePath::path` is strange... How about renaming `AzurePath` to `AzureLocation` and using `container` for `a` and `path` for `b/c/d`?

FYI:
* Azure Blob Storage calls "a" as "container name" and "b/c/d" as "blob name" in https://learn.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction .
* Azure Data Lake Storage Gen2 calls "a/b/c" as "path" and "d" as "file_name" in https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-introduction-abfs-uri .
* We don't support `abfs://...` URI yet. (`FileSystemFromUri()` doesn't accept `abfs://...` yet.)

Reply:
Oh. I misinterpreted the meaning of `path_to_file`. I think `path` would be OK. And `Azure{Path->Location}` is also a good rename.

Reply:
I've opened a new issue for this: #38758
I'll do it after we merge this.